### ISSUE TITLE
 the excess builds were not deleted

### ISSUE THREAD

**Issue by liuyp2018** [kubesphere/ks-devops#393]
**Describe the Bug**
1. The maximum number of build records is set to 1
2. In the build list, the excess builds were not deleted

![image](https://user-images.githubusercontent.com/36271543/144394758-14e37ad9-3908-4fb1-8e3f-34ed44bc35d8.png)

**Versions Used**
KubeSphere: kubesphere v3.2.1-rc.1
Kubernetes: (If KubeSphere installer used, you can skip this)

/assign @JohnNiang 
/kind bug
/priority Medium

**Comment by mzmuer** [comment]
I am interested in this issue and hope to try to solve it

**Comment by LinuxSuRen** [comment]
@mzmuer Great! Please let us know if you need any kind of help.

**Comment by mzmuer** [comment]
@LinuxSuRen @JohnNiang

**Comment by mzmuer** [comment]
@LinuxSuRen @JohnNiang 
I have found out the cause of the problem, Is this a legacy issue? 
I'm prepare to remove excess records at the end(finish, cancel, fail) of the job execution or process them when the list is returned.
May I ask what kind of practice is generally preferred according to the project.

**Comment by LinuxSuRen** [comment]
Normally, I prefer to remove the PipelineRuns in two ways:

* [pipeline gc command](https://github.com/kubesphere-sigs/ks/blob/master/kubectl-plugin/pipeline/gc.go)
* Receive the delete event from Jenkins (manual or automated), then ks-devops can delete it
  * For this case, I guess there are two components: event-receiver in ks-devops, event sender as a Jenkins plugin
  * [pipeline-event-plugin](https://github.com/linuxsuren/pipeline-event-plugin) is a draft project to play as an event sender
  * we have not started the event receiver part

**Comment by mzmuer** [comment]
> Normally, I prefer to remove the PipelineRuns in two ways:
> 
> * [pipeline gc command](https://github.com/kubesphere-sigs/ks/blob/master/kubectl-plugin/pipeline/gc.go)
> * Receive the delete event from Jenkins (manual or automated), then ks-devops can delete it
>   
>   * For this case, I guess there are two components: event-receiver in ks-devops, event sender as a Jenkins plugin
>   * [pipeline-event-plugin](https://github.com/linuxsuren/pipeline-event-plugin) is a draft project to play as an event sender
>   * we have not started the event receiver part

I see. Event-drive is a better solution. So now when the job end(or list returned) use similar [cleanPipelineRunInNamespace](https://github.com/kubesphere-sigs/ks/blob/b8bb7bcb1622c70e5ae2cf199a4975cf3bf7821f/kubectl-plugin/pipeline/gc.go#L71)  function to temporary solve the problem?

**Comment by LinuxSuRen** [comment]
Yes, I'm thinking about if you're interested in the event-driven feature. I personally prefer to have a final solution in release 3.3 instead of a temporary one.

By the way, we have a SIG meeting this week. Please join us and let's talk about it if you have free time.

**Comment by mzmuer** [comment]
> Yes, I'm thinking about if you're interested in the event-driven feature. I personally prefer to have a final solution in release 3.3 instead of a temporary one.
> 
> By the way, we have a SIG meeting this week. Please join us and let's talk about it if you have free time.

thank you, very interested and to join yous. But I'm new to DevOps, so I still play a learning role.

**Comment by LinuxSuRen** [comment]
We have a WeChat. Please add me (ID: linuxsuren) as a WeChat contact, I can let you in it.

**Comment by yudong2015** [comment]
reference issue: #847
